ext/openssl/xp_ssl.c

```diff
@@ -950,7 +951,39 @@ static int php_openssl_set_local_cert(SSL_CTX *ctx, php_stream *stream) /* {{{ *
 			}
 			GET_VER_OPT_STRING("local_pk", private_key);
 
-			if (private_key) {
+			if (private_key && (strncmp(private_key, "pkcs11:", 7) == 0)) {
+				/* local_pk is a RFC7512 URI */
+				ENGINE *engine;
+				EVP_PKEY *pkey;
+				engine = ENGINE_by_id("pkcs11");
+				if (engine == NULL) {
+					php_error_docref(NULL, E_WARNING, "Could not bind PKCS11 engine");
+					return FAILURE;
+				}
+#if 0
+				if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+					ENGINE_free(engine);
+					php_error_docref(NULL, E_WARNING, "PKCS11 engine VERBOSE error");
+					return FAILURE;
+				}
+#endif
+				if (!ENGINE_init(engine)) {
+					php_error_docref(NULL, E_WARNING, "Could not initialize PKCS11 engine");
+					return FAILURE;
+				}
+				/* ENGINE_init() returned a functional reference, so free the structural reference from ENGINE_by_id(). */
+				ENGINE_free(engine);
+				pkey = ENGINE_load_private_key(engine, private_key, 0, 0);
+				if (pkey == NULL) {
+					php_error_docref(NULL, E_WARNING, "Could not load private key %s", private_key);
+					return FAILURE;
```

Comment:
Should `ENGINE_finish` be called before returning?

Reply:
Yes, it should. Check `load_key()` from openssl source code:
https://github.com/openssl/openssl/blob/86a90dc749af91f8a7b8da6628c9ffca2bae3009/apps/apps.c#L736
```
            if (ENGINE_init(e)) {
                pkey = ENGINE_load_private_key(e, file, ui_method, &cb_data);
                ENGINE_finish(e);
```